pkg/controller/tailoredprofile/tailoredprofile_controller.go

```diff
@@ -187,6 +275,320 @@ func (r *ReconcileTailoredProfile) Reconcile(ctx context.Context, request reconc
+	var warningMessage string
+	if len(variableNeedtoBeDeprecatedList) > 0 {
+		warningMessage = fmt.Sprintf("The following variables are deprecated and need to be removed from the TailoredProfile: %s", strings.Join(variableNeedtoBeDeprecatedList, ","))
+	}
+	if len(ruleNeedToBeDeprecatedList) > 0 {
+		if warningMessage != "" {
+			warningMessage = fmt.Sprintf("%s\nThe following rules are deprecated and need to be removed from the TailoredProfile: %s", warningMessage, strings.Join(ruleNeedToBeDeprecatedList, ","))
+		} else {
+			warningMessage = fmt.Sprintf("The following rules are deprecated and need to be removed from the TailoredProfile: %s", strings.Join(ruleNeedToBeDeprecatedList, ","))
+		}
+	}
+	if len(ruleNeedToBeMigratedList) > 0 {
+		if warningMessage != "" {
+			warningMessage = fmt.Sprintf("%s\nThe following rules are migrated and need to be migrated or removed from the TailoredProfile: %s", warningMessage, strings.Join(ruleNeedToBeMigratedList, ","))
+		} else {
+			warningMessage = fmt.Sprintf("The following rules are migrated and need to be migrated or removed from the TailoredProfile: %s", strings.Join(ruleNeedToBeMigratedList, ","))
+		}
+	}
+	return warningMessage
+}
+
+// handleMigration check if there are any migrated rules in the TailoredProfile
+// and we will handle the migration of the tailored profile accordingly
+func (r *ReconcileTailoredProfile) handleMigration(
+	v1alphaTp *cmpv1alpha1.TailoredProfile, logger logr.Logger, pruneOudated bool) (doContinue bool, ruleNeedToBeMigratedList []string, err error) {
+	doContinue = true
+	// Get the list of KubeletConfig rules that are migrated with checkType change
+	migratedRules, err := r.getMigratedRules(v1alphaTp, logger)
+	if err != nil {
+		return false, nil, err
+	}
+
+	if len(migratedRules) == 0 {
+		return true, nil, nil
+	}
+
+	profileType := utils.GetScanType(v1alphaTp.GetAnnotations())
+
+	v1alphaTpCP := v1alphaTp.DeepCopy()
+
+	// check if there are any disabled rules that are migrated
+	if len(v1alphaTp.Spec.DisableRules) > 0 {
+		var newRules []cmpv1alpha1.RuleReferenceSpec
+		for ri := range v1alphaTp.Spec.DisableRules {
+			rule := &v1alphaTp.Spec.DisableRules[ri]
+			if checkType, ok := migratedRules[rule.Name]; ok && checkType != string(profileType) {
+				// remove the rule from the list of disabled rules
+				if pruneOudated {
+					doContinue = false
+					logger.Info("Removing migrated rule from disableRules", "rule", rule.Name)
+					r.Eventf(v1alphaTp, corev1.EventTypeWarning, "TailoredProfileMigratedRule", "Removing migrated rule: %s from disableRules", rule.Name)
+				} else {
+					logger.Info("Migrated rule detected in disableRules", "rule", rule.Name)
+					r.Eventf(v1alphaTp, corev1.EventTypeWarning, "TailoredProfileMigratedRule", "Migrated rule detected: %s Please migrate it or remove it from the TailoredProfile", rule.Name)
```

Comment:
Here we could be more explicit in the messaging where we saying something like "{ruleName} changed from {oldType} to {newType} and is no longer useful in TailoredProfile {tailoredProfileName}."

Reply:
that's a good point!